### PR TITLE
Start of TypeScript type definitions file and updated index.js to access trie-map.js

### PR DESCRIPTION
**Pull request by kralcifer**
Not sure if you're getting rid of Trie or not.  I had to tweak index.js to know about TrieMap.  I hope you can add this TypeScript type definitions file with next release too.  Thanks.

### CODE REVIEW

**Comment by Yomguithereal** [comment]
I have several questions:

1) Would it be possible to have a `types` folder with a definition file per structure that we would "require" in the index.d.ts.
2) What need to be done in package.json to ease Typescript consumption?

**Comment by kralcifer** [comment]
Answer for #1: I'm not sure on the types folder question since I'm new myself to making one of these.  I'll see what I can do and get back to you.

Answer for #2: Adding 'types' to package.json is the only thing and that can be omitted if we want:

From [TypeScript-DeclarationFiles-Publishing](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html)
> Also note that if your main declaration file is named index.d.ts and lives at the root of the package (next to index.js) you do not need to mark the "types" property, though it is advisable to do so.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        constructor(hashFuncs: HashFunc | HashFuncTuple, container?: any)
+        clear(): void;
+        add(item: any): FuzzyMultiMap;
+        set(key: any, item: any): FuzzyMultiMap;
+        get(key: any): any;        
+        has(item: any): boolean;
+        forEach(callback: (item: any) => any, scope: any): void;
+        values(): Iterator<any>;
+        inspect(): any;
+        [key: string]: any;
```
Comment (Yomguithereal) [inline index.d.ts]: What does this part mean?

Comment (kralcifer) [reply index.d.ts]: I was trying to put something (hopefully correct) for everything I saw in your JavaScript.
> [key: string]: any;
is an [indexable type](https://www.typescriptlang.org/docs/handbook/interfaces.html#indexable-types).  I'd say I'm not that confident in those last few and we could leave them off for now and focus on the primary methods.

Comment (kralcifer) [reply index.d.ts]: I spent a couple of hours trying various combinations of tsconfig.json, command line, npm script, command line options, etc. but didn't have any luck.  I'm stopping for the moment.  btw though it is very common to see all the types in a single file.  Just not sure if they were generated from individual or not.

**Comment by Yomguithereal** [comment]
Hello @kralcifer. I got my head around typings and started to type every structure in the `types` branch. I will close this branch now however. Thanks for your help. I'll start by typing the Trie etc. so you can have them sooner.

**Comment by Yomguithereal** [comment]
@kralcifer: just published `0.21.0`. Library is now fully typed :)

**Pull request closed without merging**